core/src/main/java/org/infinispan/transaction/impl/AbstractCacheTransaction.java

```diff
@@ -230,13 +233,37 @@ public void clearLockedKeys() {
       lockedKeys = null;
    }
 
-   private boolean hasLockOrIsLockBackup(Object key) {
-      //stopgap fix for ISPN-2728. The real fix would be to synchronize this with the intrinsic lock.
+   @Override
+   public boolean containsLockOrBackupLock(Object key) {
       Set<Object> lockedKeysCopy = lockedKeys;
       Set<Object> backupKeyLocksCopy = backupKeyLocks;
       return (lockedKeysCopy != null && lockedKeysCopy.contains(key)) || (backupKeyLocksCopy != null && backupKeyLocksCopy.contains(key));
    }
 
+   @Override
+   public boolean containsAnyLockOrBackupLock(Stream<Object> keys) {
+      return findFirstLockedOrBackupLocked(keys) != null;
+   }
+
+   @Override
+   public Object findFirstLockedOrBackupLocked(Stream<Object> keys) {
+      Set<Object> lockedKeysCopy = lockedKeys;
+      Set<Object> backupKeyLocksCopy = backupKeyLocks;
+      if (lockedKeysCopy != null && backupKeyLocksCopy != null) {
+         return keys.filter(key -> lockedKeysCopy.contains(key) || backupKeyLocksCopy.contains(key)).findAny().orElse(null);
+      } else if (lockedKeysCopy != null) {
+         return keys.filter(lockedKeysCopy::contains).findAny().orElse(null);
+      } else if (backupKeyLocksCopy != null) {
+         return keys.filter(backupKeyLocksCopy::contains).findAny().orElse(null);
+      }
+      return null;
+   }
```

Comment:
For future reference, I ran the JMH benchmark in the most voted response on
that page and the difference between streams and foreach was quite
significant:

N=10_000
Benchmark Mode Cnt Score Error Units
StreamVsVanilla.stream avgt 200 10.421 ± 0.179 ns/op
StreamVsVanilla.vanilla avgt 200 7.813 ± 0.081 ns/op

N=5
Benchmark Mode Cnt Score Error Units
StreamVsVanilla.stream avgt 200 29.490 ± 0.190 ns/op
StreamVsVanilla.vanilla avgt 200 12.885 ± 0.119 ns/op

N=1
Benchmark Mode Cnt Score Error Units
StreamVsVanilla.stream avgt 50 115.964 ± 3.237 ns/op
StreamVsVanilla.vanilla avgt 50 37.412 ± 0.712 ns/op

(N is the number of elements in the list, and the timing is per element)
Considering that Radim's SimpleCacheImpl does the entire put in 70ns
(caveat, his machine is probably a little faster than mine), I'd say it's
probably best to stay away from streams in our internals.


On Thu, Jul 23, 2015 at 3:10 PM, wburns <notifications@github.com> wrote:

> In
> core/src/main/java/org/infinispan/transaction/impl/AbstractCacheTransaction.java
> <https://github.com/infinispan/infinispan/pull/3594#discussion_r35314398>:
>
> > +      return findFirstLockedOrBackupLocked(keys) != null;
> > +   }
> > +
> > +   @Override
> > +   public Object findFirstLockedOrBackupLocked(Stream<Object> keys) {
> > +      Set<Object> lockedKeysCopy = lockedKeys;
> > +      Set<Object> backupKeyLocksCopy = backupKeyLocks;
> > +      if (lockedKeysCopy != null && backupKeyLocksCopy != null) {
> > +         return keys.filter(key -> lockedKeysCopy.contains(key) || backupKeyLocksCopy.contains(key)).findAny().orElse(null);
> > +      } else if (lockedKeysCopy != null) {
> > +         return keys.filter(lockedKeysCopy::contains).findAny().orElse(null);
> > +      } else if (backupKeyLocksCopy != null) {
> > +         return keys.filter(backupKeyLocksCopy::contains).findAny().orElse(null);
> > +      }
> > +      return null;
> > +   }
>
> This stack trace talks about it a bit more [1]. The problem is that in
> some cases it is faster and others it may not be. But in my findings most
> cases the performance difference is negligible both ways. The reason I even
> mentioned Streams in the first place to Pedro was as a way to prevent
> additional entry iterations which imo is a much bigger performance issue.
>
> [1]
> http://stackoverflow.com/questions/22658322/java-8-performance-of-streams-vs-collections#comment36791431_22670380
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/infinispan/infinispan/pull/3594/files#r35314398>.
>